ktor-client/ktor-client-cio/jvmAndNix/src/io/ktor/client/engine/cio/utils.kt

```diff
@@ -142,20 +144,21 @@ internal suspend fun writeBody(
     }
 }
 
-private suspend fun processOutgoingContent(
-    request: HttpRequestData,
-    body: OutgoingContent,
-    channel: ByteWriteChannel
-): Boolean {
+private fun needToProcessOutgoingContent(body: OutgoingContent): Boolean = when (body) {
+    is OutgoingContent.NoContent -> false
+    is OutgoingContent.ProtocolUpgrade -> throw UnsupportedContentTypeException(body)
+    is OutgoingContent.ContentWrapper -> needToProcessOutgoingContent(body.delegate())
+    else -> true
+}
```

Comment:
Might be better here to introduce an `isEmpty()` function in `OutgoingContent` then override in the implementations.  The default behaviour could be something like `contentLength?.let { it > 0 } == true`.

So then the check would be `if (request.body.isEmpty()) return`.